Rewrite "Set up development environment" (#376)
NixOS/nix.dev: source/tutorials/dev-environment.md

Let's build a Python web application using the Flask web framework as an exercise.

> For our Flask web application, create a new file called `myapp.py` and add the code:

Reviewer: ```suggestion
For our Flask web application, create a new file called `myapp.py` and add the following code:
```

@app.route("/")
def hello():
    return {
        "message": "Hello, Nix!"
    }

def run():
    app.run(host="0.0.0.0", port=5000)

if __name__ == "__main__":
    run()
```

This is a simple Flask application which serves a JSON document with the message
"Hello, Nix!".

> For the development environment, create a new file `shell.nix` which creates a
> shell environment to be used for development: 

Reviewer: That second part of the sentence was redundant.

```suggestion
To declare the development environment, create a new file `shell.nix`: 
```

```{code-block} nix shell.nix

> { pkgs ? import <nixpkgs> {} }:

Reviewer: This should use the Nixpkgs commit hash from the first pinning tutorial one encounters, otherwise the reader would have to download everything again.

pkgs.mkShell {
  packages = [
    (pkgs.python3.withPackages (ps: [
      ps.flask
    ]))

    pkgs.curl
    pkgs.jq
  ];
}
```

> This creates a shell environment with a Python 3 that contains the `flask`
> library.

Reviewer: "A Python 3" is true, but a very Nix-y way of saying things for beginners to swallow.

I think it may be better to use "package" here, even if Flask is a library, and also say `python3`, because that closely matches the example and will prevent raised eyebrows over the differences.

The wording "contains" is also ambiguous, is it the environment that contains Flask? But it also contains Python, right?

```suggestion
This creates a shell environment with `python3`, including the `flask` package.
```

requests) and [`jq`] (utility to parse and format JSON documents). Both of them
are not Python packages and they are not part of the Python ecosystem.

> If we went with Python's [virtualenv], it is not possible to have these
> utilities to be a part of the development environment without additional manual
> steps.

Reviewer: ```suggestion
If we went with Python's [virtualenv], it would not be possible to add these utilities
to the development environment without additional manual steps.
```

> requests) and `jq` (utility to parse and format JSON documents). Both of them

Reviewer: Would be cool to have links to the projects.

> If we went with Python's [virtualenv], it is not possible to have these

Reviewer: Please move the reference link below this paragraph, it's hard to find in the source otherwise.

defined in `shell.nix`.

Let's use the shell environment to start the web application:

```shell-session
[nix-shell:~/dev-environment]$ python ./myapp.py
 * Serving Flask app 'myapp'
 * Debug mode: off
WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on all addresses (0.0.0.0)
 * Running on http://127.0.0.1:5000
 * Running on http://192.168.1.100:5000
Press CTRL+C to quit

```

We have a running Python web application now, let's try it out using the
developer tools that are also included.

Launch a new terminal to start another session of the shell environment and
follow the commands below:

```shell-session
$ nix-shell

[nix-shell:~/dev-environment]$ curl 127.0.0.1:5000
{"message":"Hello, Nix!"}

[nix-shell:~/dev-environment]$ curl 127.0.0.1:5000 | jq '.message'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    26  100    26    0     0  13785      0 --:--:-- --:--:-- --:--:-- 26000
"Hello, Nix!"
```

As demonstrated, we can use both `curl` and `jq` to test the running web

> application without any manual installation, Nix just does it all for us.

Reviewer: ```suggestion
application without any manual installation.
Nix does all of that for us.
```

> With the files we created, we can add them to a repository and share them to
> other people. They can now use the same shell environment as long as they have
> Nix installed.

Reviewer: The subjects are ambiguous.

```suggestion
We can commit the files we created to version control and share them with other people.
Others can now use the same shell environment as long as they have Nix installed.
```

> 
> 

Reviewer: ```suggestion
```
